src/core/execute.c

```diff
@@ -5831,36 +5861,68 @@ int exec_spawn(Unit *unit,
 
-                r = exec_child(unit,
-                               command,
-                               context,
-                               params,
-                               runtime,
-                               cgroup_context,
-                               &exit_status);
+        f = take_fdopen(&serialization_fd, "w+");
+        if (!f)
+                return log_unit_error_errno(unit, serialization_fd, "Failed to open serialization stream: %m");
 
-                if (r < 0) {
-                        const char *status =
-                                exit_status_to_string(exit_status,
-                                                      EXIT_STATUS_LIBC | EXIT_STATUS_SYSTEMD);
+        fdset = fdset_new();
+        if (!fdset)
+                return log_oom();
 
-                        log_unit_struct_errno(unit, LOG_ERR, r,
-                                              "MESSAGE_ID=" SD_MESSAGE_SPAWN_FAILED_STR,
-                                              LOG_UNIT_INVOCATION_ID(unit),
-                                              LOG_UNIT_MESSAGE(unit, "Failed at step %s spawning %s: %m",
-                                                               status, command->path),
-                                              "EXECUTABLE=%s", command->path);
-                }
+        r = unit_serialize(unit, f, fdset, /* switching_root */ false);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize unit '%s': %m", unit->id);
 
-                _exit(exit_status);
-        }
+        r = exec_command_serialize(command, f);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecCommand: %m");
+
+        r = exec_parameters_serialize(params, f, fdset);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecParameters: %m");
+
+        r = exec_runtime_serialize(runtime, f, fdset);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecRuntime: %m");
+
+        r = exec_context_serialize(context, f, fdset);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecContext: %m");
+
+        r = exec_cgroup_context_serialize(cgroup_context, f);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize CGroupContext: %m");
+
+        if (fseeko(f, 0, SEEK_SET) == (off_t) -1)
+                return log_unit_error_errno(unit, errno, "Failed to reseek on serialization stream: %m");
+
+        r = fd_cloexec(fileno(f), false);
+        if (r < 0)
+                return log_unit_error_errno(unit, errno, "Failed to set O_CLOEXEC on serialization fd: %m");
+
+        r = fdset_cloexec(fdset, false);
+        if (r < 0)
+                return log_unit_error_errno(unit, errno, "Failed to set O_CLOEXEC on serialized fds: %m");
```

Comment:
hmm, do we have to do this in the parent? it should be equally fast or slow if we do it in the child, no? but if we do it in the child, then the parent will never have an fd with O_CLOEXEC off in its fd table, and that is a *good* thing

Reply:
I would rather keep the child callback really as minimal as possible - what would that cover against? The set is freed on return anyway, and we know we are not going to exec something else

Reply:
we generally have the rule that in our processes we only have fds open that have O_CLOEXEC set to that we don't accidentally pass any fds to the wrong processes. in fact, this is crucial because otherwise any fds you are passing to a service A and thus turn O_CLOEXEC off for will also end up being inherited into service B you fork right after. By never have the fds set to non-O_CLOEXEC in the parent (and only in the child) you avoid any such mess.

Reply:
they are closed when returning from the function, no? this is a special case, it's really critical to do the bare minimum from the child

Reply:
Are they closed? I dont think they necessarily are. And its generally a matter of hygiene to never have non-o_cloexec fds around. Also i dont see why we cant just call the bunch of fcntl is the child instead of the pare n. See no reason whatsoever...

Reply:
yes, they are closed by fdset_free

Reply:
also isn't it the same fd in the parent and child before exec due to clone_vm?

Reply:
CLONE_VM is not CLONE_FILES